### ISSUE TITLE
Question: how to abort a script?

### ISSUE THREAD

**Issue by wsw70** [custom-components/pyscript#140]
I have the following code at the very top of a script:

```
devices = {}
try:
    for device in pyscript.config['meross']:
        → here 'devices' is filled in with compulsory info from the configuration, if it is missing the whole script should crash
except KeyError as e:
    log.critical(f"missing required parameter in configuration.yaml: {e}")
    → I would like to abort here
```

Is it possible to abort a script at some point, so that what follows is not sourced (and therefore the automation or whatever does not work)?

**Comment by dlashua** [comment]
There are a few ways, but this is my favorite:

def thing_to_be_aborted():
  task.sleep(1000000)

running_task = task.create(thing_to_be_aborted)
task.sleep(5)
running_task.cancel()

**Comment by dlashua** [comment]
The other way uses `task.unique`:

```
@state_trigger('input_boolean.test_1 == "on"')
def thing_to_be_aborted():
  task.unique('running_task')
  task.sleep(10000000)

@state_trigger('input_boolean.test_1 == "on"')
def aborting_task():
  task.sleep(5)
  task.unique('running_task')
```

**Comment by craigbarratt** [comment]
The current task can kill itself using:
```
import asyncio

    ...
    task.cancel(asyncio.current_task())
    time.sleep(10000)
```
Since the canceling is done by sending a message to another task, it means statements after the `task.cancel` could still get executed, which is why the `task.sleep` call is there.

We could make `task.cancel` default to that if it is called with no arguments, and avoid the need for the user to call `task.sleep`.  Thoghts?

**Comment by dlashua** [comment]
your commit regarding `task.cancel` makes sense. 

However, I realize now that I misinterpreted the original question. The `task.sleep()` I was using was just simulating some other task being done. But, what @wsw70  want's to do is cancel the task from inside of itself.

For this, yes, this new `task.cancel()` without arguments will work perfectly. Prior to this, what I would have done is to move all of the code that is being run into a function with `@time_trigger('startup')` decorating it. Then, I can simply 'return' if I don't want the rest of the code to run.

**Comment by wsw70** [comment]
> what @wsw70 want's to do is cancel the task from inside of itself.

Sorry for not having been clear.

I am not sure what you both mean by "task", what I wanted to do is to abort the whole script (the python script). The code I mentioned is at the very top of the file and its successful execution is compulsory for the (decorated) functions (tasks?) to make sense. 

So if that piece of the code fails, the whole script should abort, and the automations just not be processed.

**Comment by craigbarratt** [comment]
Ah I see.  You could just raise the same exception in the exception handler after printing the error message:
```python
devices = {}
try:
    for device in pyscript.config['meross']:
        → here 'devices' is filled in with compulsory info from the configuration, if it is missing the whole script should crash
except KeyError as e:
    log.critical(f"missing required parameter in configuration.yaml: {e}")
    raise e
```

**Comment by wsw70** [comment]
> You could just raise the same exception in the exception handler after printing the error message

Thanks. I would like to abort more cleanly (no Traceback necessary, a critical log would be enough). I do not know how the scripts (files) are handled: would a `sys.exit()` work as well?

**Comment by dlashua** [comment]
It's important to remember that you don't want to do too much processing in the "main" scope of any pyscript. This is because the "main" flow is processed when pyscript starts up/reloads and it's all handled in one giant task. By using startup functions, you defer those tasks and allow them to run in parallel. Example:

```
if sun.sun == 'below_horizon':
  raise Exception('do not run the rest of this script')

@state_trigger('input_boolean.test_1 == "on"')
def something():
  pass
```

Without the exception, the rest of this file will run, registering the state_trigger. Additionally, if code placed where the Exception is takes time to do, pyscript startup is slowed down waiting for it. Consider this alternative:

```
rt = []

@time_trigger('startup')
def startup():
  if sun.sun == 'below_horizon':
    log.error('do not run the rest of this script')
    return

  @state_trigger('input_boolean.test_1 == "on"')
  def something():
    pass

  rt.append(something)
```

In this example, an Exception is not needed as a `return` will suffice. Additionally, because this is wrapped in a "startup" `time_trigger` all of your other pyscripts will load and then this "startup" will be handled.

**Comment by wsw70** [comment]
@dlashua Thanks Daniel for the information.

The initial task I mentioned in my question is extremely short (this is literally the creation of a `dict` filled in with information from the `pyscript` configuration. It will be a blast.

All the rest of the code are functions (decorated or not)?

This said, your example is very useful for some other scripts I have, where the processing time is not gargantuan either, but your code is a nice best practice.
